src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2370 @@
+ * and deliver the event. Otherwise just deliver the event.
+ */
+void
+circpad_deliver_sent_relay_cell_events(circuit_t *circ,
+                                       uint8_t relay_command)
+{
+  /* Padding negotiate cells are ignored by the state machines
+   * for simplicity. */
+  if (relay_command == RELAY_COMMAND_PADDING_NEGOTIATE ||
+      relay_command == RELAY_COMMAND_PADDING_NEGOTIATED) {
+    return;
+  }
+
+  /* RELAY_COMMAND_DROP is the multi-hop (aka circuit-level) padding cell in
+   * tor. (CELL_PADDING is a channel-level padding cell, which is not relayed
+   * or processed here) */
+  if (relay_command == RELAY_COMMAND_DROP) {
+    /* Optimization: The event for RELAY_COMMAND_DROP is sent directly
+     * from circpad_send_padding_cell_for_callback(). This is to avoid
+     * putting a cell_t and a relay_header_t on the stack repeatedly
+     * if we decide to send a long train of padidng cells back-to-back
+     * with 0 delay. So we do nothing here. */
+    return;
+  } else {
+    /* This is a non-padding cell sent from the client or from
+     * this node. */
+    circpad_cell_event_nonpadding_sent(circ);
+  }
+}
+
+/**
+ * Free the machineinfo at an index
+ */
+static void
+circpad_circuit_machineinfo_free_idx(circuit_t *circ, int idx)
+{
+  if (circ->padding_info[idx]) {
+    tor_free(circ->padding_info[idx]->histogram);
+    timer_free(circ->padding_info[idx]->padding_timer);
+    tor_free(circ->padding_info[idx]);
+  }
+}
+
+/**
+ * Free all padding machines and mutable info associated with
+ * circuit
+ */
+void
+circpad_circuit_machineinfo_free(circuit_t *circ)
+{
+  FOR_EACH_CIRCUIT_MACHINE_BEGIN(i) {
+    circpad_circuit_machineinfo_free_idx(circ, i);
+  } FOR_EACH_CIRCUIT_MACHINE_END;
+}
+
+/**
+ * Allocate a new mutable machineinfo structure.
+ */
+circpad_machineinfo_t *
+circpad_circuit_machineinfo_new(circuit_t *on_circ, int machine_index)
+{
+  circpad_machineinfo_t *mi = tor_malloc_zero(sizeof(circpad_machineinfo_t));
+  mi->machine_index = machine_index;
+  mi->on_circ = on_circ;
+
+  return mi;
+}
+
+static void
+circpad_setup_machine_on_circ(circuit_t *on_circ, circpad_machine_t *machine)
+{
+  if (CIRCUIT_IS_ORIGIN(on_circ) && !machine->origin_side) {
+    log_fn(LOG_WARN, LD_BUG,
+           "Can't set up non-origin machine on origin circuit!");
+    return;
+  }
+
+  if (!CIRCUIT_IS_ORIGIN(on_circ) && machine->origin_side) {
+    log_fn(LOG_WARN, LD_BUG,
+           "Can't set up origin machine on non-origin circuit!");
+    return;
+  }
+
+  tor_assert_nonfatal(on_circ->padding_machine[machine->machine_index]
+                      == NULL);
+  tor_assert_nonfatal(on_circ->padding_info[machine->machine_index] == NULL);
+
+  on_circ->padding_info[machine->machine_index] =
+      circpad_circuit_machineinfo_new(on_circ, machine->machine_index);
+  on_circ->padding_machine[machine->machine_index] = machine;
+}
+
+static void
+circpad_circ_client_machine_init(void)
```

Comment:
So we can do this a few different ways.. We can set time limits that mark a circuit as not to be closed until the timer is up, or we can set a flag on machines that say don't close the circuit if this machine is active. Or we could set that flag, and have an additional timer that says the machine stays alive until its up.

Timers are hard because we'd need a distribution that specifies lifetime. Setting a flag that says "don't close circuit if this machine is active" is easy. And then those machines can shut themselves down using state length limits, infinity bin probability, and or bins empty.

What do you think about these options?

Reply:
I lilke the approach of a machine flag that says" don't close circuit if this machine is active". I think that makes sense and it would work in the case of IP circs.

Reply:
OK opened a ticket for the "dont close circ if machine is active" idea: https://trac.torproject.org/projects/tor/ticket/28780

So I think we can perhaps aim for the following to get a useful circ machine for 040 (unclear if we will have enough time tho):

- Do both tickets #28633 and #28780.
- For circuit purposes `CIRCUIT_PURPOSE_C_INTRO*` we introduce a simple padding machine that does some light randomized padding to obfuscate the introduction activity. We also setup the #28780 flag to not close the circuit before the machine is done.
- For circuit purposes `CIRCUIT_PURPOSE_C_*REND*` we introduce another simple padding machine that does some light randomized padding to obfuscate rendezvous setup/waiting.

This way we will slightly obfuscate the introduction and rendezvous circuits. The other remaining thing we could do is obfuscate the HS descriptor fetch but that might be harder, so perhaps we can skip this for 040.

How does that sound to you, Mike?

Reply:
This sounds reasonable. I believe that #28633 is supported already in the current code. #28780 will be a little tricky since we need to go through the hs code and circuit idle timeout code and add checks that say "if there's a machine here, don't close the circuit" and still remember to close the circuit once the machine is gone (which might be doable from the circuit idle handling).